src/main/java/io/pravega/connectors/flink/FlinkPravegaWriter.java

```diff
@@ -160,12 +160,12 @@ public void invoke(T event) throws Exception {
         }
 
         this.pendingWritesCount.incrementAndGet();
-        final AckFuture ackFuture = this.pravegaWriter.writeEvent(this.eventRouter.getRoutingKey(event), event);
+        final CompletableFuture future = this.pravegaWriter.writeEvent(this.eventRouter.getRoutingKey(event), event);
         if (writerMode == PravegaWriterMode.ATLEAST_ONCE) {
-            ackFuture.addListener(
+            future.thenRunAsync(
```

Comment:
I am not 100% sure this is the right replacement but it seemed like it might be.

Reply:
I don't think it is because `thenRunAsync` runs only if the future completed normally, not exceptionally.   Meanwhile the catch block seems intended to handle an exception from `writeEvent`. All told, `whenCompleteAsync` makes the most sense here.

Pseudo-code:
```
future.whenCompleteAsync(
                    (result, e) -> {
                        if(e == null) {
                            synchronized (this) {
                                pendingWritesCount.decrementAndGet();
                                this.notify();
                            }
                        }
                        else {
                            log.warn("Detected a write failure: {}", e);

                            // We will record only the first error detected, since this will mostly likely help with
                            // finding the root cause. Storing all errors will not be feasible.
                            writeError.compareAndSet(null, e);
                        }
                    },
                    executorService
            );
```